### DISCUSSION TITLE
Testing: Request-Object in Illuminate\Testing\TestResponse

### DISCUSSION THREAD

**Discussion by tduesterhus** [laravel/framework#46499] [Ideas]
Hi all,

I started to test our OpenAPI docs against the requests/responses in our feature tests to make sure they are in sync with the actual implementation. But it turned out that it is impossible to get the request generated in _Illuminate\Foundation\Testing\Concerns\MakesHttpRequests::call()_. So I had to write a workaround to overwrite the _call_-function. From my point of view, the better solution would be to add the Request to the _Illuminate\Testing\TestResponse_-class so that it could be used for testing if required.
What do you think?

Best,
Tobias

**Comment by joelclermont** [comment]
Accessing the original request via the `TestResponse` could work, but it would also be useful to have the ability generate the request using the `MakesHttpRequests` helpers without actually simulating a `call`. I wouldn't even issue the call in my test if the constructed request failed the OpenAPI spec validation.

I'm envisioning something like this:
```php
    public function testIndexSuccess(): void
    {
        $this->createActingAs();

        $request = $this->getJsonRequest(route('api.v1.words.index')); // each helper could have a ...Request version that just builds and returns the request
        $this->assertMatchesAPIV2Spec($request);

        $response = $this->callRequest($request); // and we'd need a new single helper to call that request object
        $this->assertMatchesAPIV2Spec($response);

        $response->assertOk();
        $response->assertJsonFragment([ /* not relevant... */]);
    }
```

**Comment by tduesterhus** [comment]
Hi @joelclermont,

good point, this would be even better. Cleaner and more flexible 👍
